tutorials/flappy-bird/step_8/flappy-bird.cpp

```diff
@@ -0,0 +1,660 @@
+
+        return entity;
+    }
+
+    // Factory for pipes, requires to know if it's a top one, position x of the column, and the gap starting position Y
+    pipe create_pipe(entt::registry &registry, bool is_top, float pos_x, float gap_start_pos_y) {
+        // Retrieve constants
+        const auto canvas_height = registry.ctx<graphics::canvas_2d>().canvas.size.y();
+        const auto constants = registry.ctx<flappy_bird_constants>();
+
+        // PIPE BODY
+        // Top pipe is at Y: 0 and bottom pipe is at canvas_height, bottom of the canvas
+        transform::position_2d body_pos{pos_x, is_top ? 0.f : canvas_height};
+
+        // Size X is the column thickness,
+        // Size Y is the important part.
+        // If it's a top pipe, gap_start_pos_y should be bottom of the rectangle
+        //  So half size should be gap_start_pos_y since center of the rectangle is at 0.
+        // If it's the bottom pipe, top of the rectangle will be at gap_start_pos_y + gap_height
+        //  So half size should be canvas_height - (gap_start_pos_y + gap_height)
+        // Since these are half-sizes, and the position is at the screen border, we multiply these sizes by two
+        math::vec2f body_size{constants.column_thickness,
+                              is_top ?
+                              gap_start_pos_y * 2.0f :
+                              (canvas_height - (gap_start_pos_y + constants.gap_height)) * 2.0f};
+
+        auto body = geometry::blueprint_rectangle(registry, body_size, constants.pipe_color, body_pos,
+                                                  constants.pipe_outline_color);
+
+        // PIPE CAP
+        // Let's prepare the pipe cap
+        // Size of the cap is defined in constants
+        math::vec2f cap_size{constants.column_thickness + constants.pipe_cap_extra_width, constants.pipe_cap_height};
+
+        // Position, X is same as the body. Bottom of the cap is aligned with bottom of the body,
+        // or start of the gap, we will use start of the gap here, minus half of the cap height
+        transform::position_2d cap_pos{body_pos.x(),
+                                       is_top ?
+                                       gap_start_pos_y - constants.pipe_cap_height * 0.5f :
+                                       gap_start_pos_y + constants.gap_height + constants.pipe_cap_height * 0.5f
+        };
+
+        // Construct the cap
+        auto cap = geometry::blueprint_rectangle(registry, cap_size, constants.pipe_color, cap_pos,
+                                                 constants.pipe_outline_color);
+
+        // Set layers, cap should be in front of body
+        registry.assign<graphics::layer<4>>(cap);
+        registry.assign<graphics::layer<3>>(body);
+        registry.assign<entt::tag<"game_scene"_hs>>(cap);
+        registry.assign<entt::tag<"game_scene"_hs>>(body);
+        registry.assign<entt::tag<"dynamic"_hs>>(cap);
+        registry.assign<entt::tag<"dynamic"_hs>>(body);
```

Comment:
We are repeating ourselves here.

We can an utility function in the anonymous namespace at the top of the file that take an entity and add the tags according to your needs

`add_game_dynamic_entity(cap)` and internally set the tags.

if you are lazy you can also do:

```cpp
auto fill_dynamic_entity_functor = [this](entt::entity entity) {
   registry.assign<entt::tag<"game_scene"_hs>>(entity);
   registry.assign<entt::tag<"dynamic"_hs>>(entity);
}

fill_dynamic_entity_functor(cap);
fill_dynamic_entity_functor(body);
```